### PR TITLE
Sahar's JSON + V1 studio ui

### PR DESCRIPTION
**Pull request by caesar2164**
New, new PR.

That old pull request got super cluttered.

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    ],
+    install_requires=[
+        'XBlock',
+    ],
+    entry_points={
+        'xblock.v1': [
+            'platformtour = platformtour:PlatformTourXBlock',
+        ]
+    },
+    package_data=package_data("platformtour", ["static", "public"]),
```
Comment (stvstnfrd) [inline setup.py]: I think you'll need to add 'templates' here?

File: setup.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+            for fname in files:
+                data.append(os.path.relpath(os.path.join(dirname, fname), pkg))
+
+    return {pkg: data}
+
+
+setup(
+    name='platform-tour-xblock',
+    version='0.1',
+    description='Platform Tour XBlock',   # TODO: write a better description.
```
Comment (stvstnfrd) [inline setup.py]: Do the TODO?

File: setup.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    for root in roots:
+        for dirname, _, files in os.walk(os.path.join(pkg, root)):
+            for fname in files:
+                data.append(os.path.relpath(os.path.join(dirname, fname), pkg))
+
+    return {pkg: data}
+
+
+setup(
+    name='platform-tour-xblock',
```
Comment (stvstnfrd) [inline setup.py]: Can we call this `xblock-platform-tour` instead?
It sorts better..

File: setup.py [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    version='0.1',
```
Comment (stvstnfrd) [inline setup.py]: Should we preëmptively set this to `1.0` since whatever we merge here will end up as the first public release?

File: setup.py [hunk 5]
```diff
@@ -0,0 +1,41 @@
+"""Setup for Platform Tour XBlock."""
```
Comment (stvstnfrd) [inline setup.py]: Multiline docstring, per style guide, please.

File: setup.py (hunk 1 above)
Comment (stvstnfrd) [reply setup.py]: Also, single quotes..

File: setup.py [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+
+import os
+
+from setuptools import setup
+
+
+def package_data(pkg, roots):
+    """Generic function to find package_data.
```
Comment (stvstnfrd) [inline setup.py]: Multiline docstring formatting..

File: setup.py [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+    All of the files under each of the `roots` will be declared as package
+    data for package `pkg`.
+
+    """
+    data = []
+    for root in roots:
+        for dirname, _, files in os.walk(os.path.join(pkg, root)):
+            for fname in files:
+                data.append(os.path.relpath(os.path.join(dirname, fname), pkg))
+
```
Comment (stvstnfrd) [inline setup.py]: Nit: newline.

File: platformtour/default_steps.py [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+        "selector": "a.doc-link",
+        "dataIntro": "For any technical issues or platform-specific questions, click on the \"Help\" link to access the Help Center or contact support.",
+        "dataPosition": "bottom",
+    },
+    {
+        "key": "tour_done",
+        "title": "End of Platform Tour",
+        "selector": "div.course-wrapper",
+        "dataIntro": "That concludes the platform tour. \n\n Click Done to close this walkthrough.",
+        "dataPosition": "top",
```
Comment (stvstnfrd) [inline platformtour/default_steps.py]: Single quotes throughout...

File: platformtour/default_steps.py [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+]
+
+def get_choices(keys):
+    choices_list = []
+    for step in DEFAULT_STEPS:
+        key = step.get('key')
+        title = step.get('title')
+        if key and title:
+            is_choice_enabled = key in keys
+            choice = {"key": key, "title": title, "enabled": is_choice_enabled}
```
Comment (stvstnfrd) [inline platformtour/default_steps.py]: Make this multiline and use single quotes.

File: platformtour/default_steps.py [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+            choices_list.append(choice)
+    is_custom_enabled = "custom" in keys
+    choices_list.append({"key": "custom", "title": "Custom (Advanced Users)", "enabled": is_custom_enabled})
+    return choices_list
+
+def get_default_keys():
+    keys_list = []
```
Comment (stvstnfrd) [inline platformtour/default_steps.py]: This is more Pythonic as a list comprehension:

```py
keys = [
    step.get('key')
    for step in DEFAULT_STEPS
]
return keys
```

File: platformtour/default_steps.py [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+    for step in DEFAULT_STEPS:
+        keys_list.append(step.get('key'))
+    return keys_list
+
+def get_display_steps(keys):
+    step_list = []
```
Comment (stvstnfrd) [inline platformtour/default_steps.py]: This is more Pythonic as a list comprehension:

```py
steps = [
    step
    for step in DEFAULT_STEPS
    if step.get('key') in keys
]
return steps
```

File: platformtour/default_steps.py [hunk 12]
```diff
@@ -0,0 +1,144 @@
...
+    choices_list = []
+    for step in DEFAULT_STEPS:
+        key = step.get('key')
+        title = step.get('title')
+        if key and title:
+            is_choice_enabled = key in keys
+            choice = {"key": key, "title": title, "enabled": is_choice_enabled}
+            choices_list.append(choice)
+    is_custom_enabled = "custom" in keys
+    choices_list.append({"key": "custom", "title": "Custom (Advanced Users)", "enabled": is_custom_enabled})
```
Comment (stvstnfrd) [inline platformtour/default_steps.py]: Make this multiline and use single quotes.

File: platformtour/default_steps.py [hunk 13]
```diff
@@ -0,0 +1,144 @@
...
+def get_choices(keys):
+    choices_list = []
+    for step in DEFAULT_STEPS:
+        key = step.get('key')
+        title = step.get('title')
+        if key and title:
+            is_choice_enabled = key in keys
+            choice = {"key": key, "title": title, "enabled": is_choice_enabled}
+            choices_list.append(choice)
+    is_custom_enabled = "custom" in keys
```
Comment (stvstnfrd) [inline platformtour/default_steps.py]: Single quotes.

File: platformtour/platformtour.py [hunk 14]
```diff
@@ -0,0 +1,170 @@
+"""
+This is the core logic for the Platform Tour xblock, which introduces students to
+a course through a digital tour.
+"""
+import json
+import pkg_resources
+from django.template import Context
```
Comment (stvstnfrd) [inline platformtour/platformtour.py]: Newline between `pkg_resources` and `django`?

File: platformtour/platformtour.py [hunk 15]
```diff
@@ -0,0 +1,170 @@
...
+from django.template.loader import get_template
+from xblock.core import XBlock
+from xblock.fields import List, Scope, String
+from xblock.fragment import Fragment
+
+import default_steps
```
Comment (stvstnfrd) [inline platformtour/platformtour.py]: Don't treat this like a global.
Do `from platformtour import default_steps` or whatever..

File: platformtour/platformtour.py [hunk 16]
```diff
@@ -0,0 +1,170 @@
+"""
+This is the core logic for the Platform Tour xblock, which introduces students to
+a course through a digital tour.
+"""
+import json
+import pkg_resources
+from django.template import Context
+from django.template.loader import get_template
+from xblock.core import XBlock
+from xblock.fields import List, Scope, String
```
Comment (stvstnfrd) [inline platformtour/platformtour.py]: Multiple lines..

File: platformtour/platformtour.py [hunk 17]
```diff
@@ -0,0 +1,170 @@
...
+from xblock.fragment import Fragment
+
+import default_steps
+
+def _resource_string(path):
+    """Handy helper for getting resources from our kit."""
```
Comment (stvstnfrd) [inline platformtour/platformtour.py]: Multiline..

File: platformtour/platformtour.py [hunk 18]
```diff
@@ -0,0 +1,170 @@
...
+    button_label = String(
+        display_name=('Button label'),
+        help=('This is the text that will appear on the button'),
+        default='Begin Platform Tour',
+        scope=Scope.settings,
+    )
+
+    intro = String(
+        display_name=('Introduction text'),
+        help=('This is the introduction that will precede the button'
```
Comment (stvstnfrd) [inline platformtour/platformtour.py]: Break this up better:
```py
help=(
    'This is ... '
    'more text.'
)
```

File: platformtour/platformtour.py [hunk 19]
```diff
@@ -0,0 +1,170 @@
...
+            ' and explain its presence to the user'),
+        default='Click the button below to learn how to navigate the platform!',
```
Comment (stvstnfrd) [inline platformtour/platformtour.py]: Is the exclamation mark necessary? ;)

File: platformtour/platformtour.py [hunk 20]
```diff
@@ -0,0 +1,170 @@
...
+
+    step_choices = List(
+        display_name=('Choose the steps for the Platform Tour'),
+        help=('List representing steps of the tour'),
+        default=[],
+        multiline_editor=True,
+        scope=Scope.settings,
+        resettable_editor=False,
+    )
+
```
Comment (stvstnfrd) [inline platformtour/platformtour.py]: Nitpick, but I like removing the empty lines between field declarations.
The indents and parens set them apart from one another, while letting them be a bit more closely displayed.

File: platformtour/platformtour.py [hunk 21]
```diff
@@ -0,0 +1,170 @@
...
+        scope=Scope.settings,
+    )
+
+    def build_fragment(
+        self,
+        template,
+        contect_dict,
+    ):
+        context = Context(contect_dict)
+        fragment = Fragment(template.render(context))
```
Comment (stvstnfrd) [inline platformtour/platformtour.py]: We should look at the PRs we're getting that use the newer style of django template loading and if that's better, we should do it that way.

File: platformtour/platformtour.py [hunk 22]
```diff
@@ -0,0 +1,170 @@
...
+                'intro': self.intro,
+                'steps': steps,
+            }
+        )
+        template = get_template('platformtour.html')
+        fragment = self.build_fragment(
+            template,
+            context
+        )
+        fragment.add_css(_resource_string('static/css/platformtour.css'),)
```
Comment (stvstnfrd) [inline platformtour/platformtour.py]: You want to add the URL, not the string..

File: platformtour/platformtour.py [hunk 23]
```diff
@@ -0,0 +1,170 @@
...
+        help=(
+            'This is the title for this component'
+        ),
+        default='Platform Tour',
+        scope=Scope.settings,
+    )
+
+    button_label = String(
+        display_name=('Button label'),
+        help=('This is the text that will appear on the button'),
```
Comment (stvstnfrd) [inline platformtour/platformtour.py]: But what is "the button"?

File: platformtour/platformtour.py [hunk 24]
```diff
@@ -0,0 +1,170 @@
...
+
+    intro = String(
+        display_name=('Introduction text'),
+        help=('This is the introduction that will precede the button'
+            ' and explain its presence to the user'),
+        default='Click the button below to learn how to navigate the platform!',
+        scope=Scope.settings,
+    )
+
+    step_choices = List(
```
Comment (stvstnfrd) [inline platformtour/platformtour.py]: `enabled_default_steps` or `chosen_default_steps` maybe, to be a bit more explicit?

File: platformtour/platformtour.py [hunk 25]
```diff
@@ -0,0 +1,170 @@
...
+    def workbench_scenarios():
+        """A canned scenario for display in the workbench."""
+        return [
+            ("PlatformTourXBlock",
+             """<platformtour/>
+             """),
+            ("Multiple PlatformTourXBlock",
+             """<vertical_demo>
+                <platformtour/>
+                <platformtour/>
```
Comment (stvstnfrd) [inline platformtour/platformtour.py]: You should configure some of these with other options.

File: platformtour/static/css/platformtour_studio.css [hunk 26]
```diff
@@ -0,0 +1,38 @@
+#platformtour_studio_step_choices {
```
Comment (stvstnfrd) [inline platformtour/static/css/platformtour_studio.css]: Is there a common parent selector/wrapper that we can use to prefix all of these selectors?
Similar to what we usually do on the LMS..

Comment (stvstnfrd) [reply platformtour/static/css/platformtour_studio.css]: `.xblock_platformtour_studio`?

File: platformtour/static/js/src/platformtour.js [hunk 27]
```diff
@@ -0,0 +1,24 @@
...
+            platformTourData[i].attr('data-intro', step.dataIntro);
+            platformTourData[i].attr('data-position', step.dataPosition);
+            var screenReaderListItem = '<li class="step">' + step.dataIntro + '</li>';
+            $('.screen-reader-steps', element).append(screenReaderListItem);
+            stepsAdded++;
+        }
+    });
+
+    $('button.navmaker', element).click(function(){
+        introJs().start();
```
Comment (stvstnfrd) [inline platformtour/static/js/src/platformtour.js]: It's a global call?
What happens if you have multiple walkthroughs on a single page?

File: platformtour/static/js/src/platformtour_studio.js [hunk 28]
```diff
@@ -0,0 +1,63 @@
+function PlatformTourStudioUI(runtime, element) {
+    'use strict';
+
+    var $ = window.$;
+    var $element = $(element);
+    var buttonSave = $element.find('.save-button');
+    var buttonCancel = $element.find('.cancel-button');
+    var url = runtime.handlerUrl(element, 'studio_view_save');
+
+    var customSteps = $("#platformtour_studio_custom_steps");
```
Comment (stvstnfrd) [inline platformtour/static/js/src/platformtour_studio.js]: Can you just give the element a classname like custom-steps and then lookup with:

`$element.find('.custom-steps')` ?

File: platformtour/static/js/src/platformtour_studio.js [hunk 29]
```diff
@@ -0,0 +1,63 @@
...
+    var stepChoices = $("#platformtour_studio_step_choices");
+
+    var checkboxIconCheckedClass = "fa-check-square-o ";
+    var checkboxIconUncheckedClass = "fa-square-o ";
+
+    $(".step input", stepChoices).click(function() {
+        var stepSelected = $(this).attr('name');
+        $(this).siblings("i").toggleClass(checkboxIconCheckedClass + checkboxIconUncheckedClass);
+        if (stepSelected == "custom") {
```
Comment (stvstnfrd) [inline platformtour/static/js/src/platformtour_studio.js]: Triple equal?

**Review by stvstnfrd — CHANGES_REQUESTED** [review]
Here's a first pass.

I stopped mentioning it at a point, but be sure to use single quotes for non-interpolated strings, instead of switching back and forth.

I think we can get rid of most, if not all of the HTML IDs we're currently using;
I've included some suggestions.

We should also cross-reference this with the community PRs we've received;
let's make sure this works in Hawthorn too, out of the box.

File: platformtour/platformtour.py (hunk 19 above)
Comment (caesar2164) [reply platformtour/platformtour.py]: that's what I was given as text, but I can remove it, lol

File: platformtour/platformtour.py (hunk 24 above)
Comment (caesar2164) [reply platformtour/platformtour.py]: how about `chosen_steps`?

File: platformtour/platformtour.py (hunk 15 above)
Comment (caesar2164) [reply platformtour/platformtour.py]: is that going to work if `default_steps.py` is a sibling of this file?

File: platformtour/static/js/src/platformtour_studio.js (hunk 28 above)
Comment (caesar2164) [reply platformtour/static/js/src/platformtour_studio.js]: Studio uses IDs on all entry fields. So I don't think changing this is necessary? See `platformtour_studio.html`, the first fields are literal copies of what Studio generated with the `studio_editable` mixin...

File: platformtour/static/css/platformtour_studio.css (hunk 26 above)
Comment (caesar2164) [reply platformtour/static/css/platformtour_studio.css]: See my reply in `platformtour_studio.js`,

File: platformtour/platformtour.py (hunk 14 above)
Comment (caesar2164) [reply platformtour/platformtour.py]: done

Comment (stvstnfrd) [reply platformtour/platformtour.py]: Test.

File: platformtour/platformtour.py (hunk 15 above)
Comment (stvstnfrd) [reply platformtour/platformtour.py]: Yes it's fine.

**Review by stvstnfrd — COMMENTED** [review]
This is a test...

File: platformtour/static/js/src/platformtour_studio.js (hunk 28 above)
Comment (caesar2164) [reply platformtour/static/js/src/platformtour_studio.js]: (basically studio only ever has one xblock open for editing at a time, so IDs are ok...)

File: platformtour/static/js/src/platformtour.js (hunk 27 above)
Comment (caesar2164) [reply platformtour/static/js/src/platformtour.js]: Currently I believe the list of steps gets built by the data elements on them. (Though intro.js has the possibility for elements to be passed to it, so I can looking into that?)

File: platformtour/platformtour.py (hunk 15 above)
Comment (caesar2164) [reply platformtour/platformtour.py]: Making this not a global was not working. I did it like you showed:
```python
from platformtour import default_steps
```

File: platformtour/platformtour.py (hunk 25 above)
Comment (caesar2164) [reply platformtour/platformtour.py]: Still need to do this, though it may not be a good idea to have multiple right now...

**Review by caesar2164 — COMMENTED** [review]
I think this xBlock is ready for round 2 of review.

File: platformtour/platformtour.py [hunk 30]
```diff
@@ -0,0 +1,205 @@
...
+
+from django.template.context import Context
+from xblock.core import XBlock
+from xblock.fields import List
+from xblock.fields import Scope
+from xblock.fields import String
+from xblock.fragment import Fragment
+from xblockutils.resources import ResourceLoader
+
+from . import default_steps
```
Comment (caesar2164) [inline platformtour/platformtour.py:16]: this seems to work now!

File: platformtour/templates/platformtour_studio.html [hunk 31]
```diff
@@ -0,0 +1,61 @@
+<div class="xblock_platformtour_studio wrapper-comp-settings editor-with-buttons is-active" id="settings-tab">
+    <ul class="list-input settings-list">
+         <li class="field comp-setting-entry is-set">
+            <div class="wrapper-comp-setting">
+                <label class="label setting-label" for="platformtour_studio_display_name">Display Name</label>
+                <input class="input setting-input" id="platformtour_studio_display_name" value="{{ display_name }}" type="text">
```
Comment (caesar2164) [inline platformtour/templates/platformtour_studio.html]: these have to be IDs for the `for` attributes of the labels to connect to them correctly. (Also if you implicit label, styling breaks)

File: platformtour/public/js/src/platformtour.js [hunk 32]
```diff
@@ -0,0 +1,8 @@
+function PlatformTourXBlock(runtime, element) {
+    $('button.navmaker', element).click(function(){
+        var stepsJson = JSON.parse($('.platformtour-steps', element).text());
+        var intro = introJs();
+        intro.setOptions({'steps': stepsJson});
```
Comment (caesar2164) [inline platformtour/public/js/src/platformtour.js:5]: now using programmatic JSON intro.js, better and improved!

File: platformtour/platformtour.py [hunk 33]
```diff
@@ -0,0 +1,205 @@
...
+        path = os.path.join('public', path)
+        resource_url = self.runtime.local_resource_url(self, path)
+        return resource_url
+
+    def build_fragment(
+        self,
+        rendered_template,
+        initialize_js_func,
+        additional_css=[],
+        additional_js=[],
```
Comment (stvstnfrd) [inline platformtour/platformtour.py]: Pylint should be complaining about dangerous defaults.

You'll want to use the form like:

```py
def build_fragment(additional_js=None):
    additional_js = additional_js or []
```

File: platformtour/platformtour.py (hunk 25 above)
Comment (stvstnfrd) [reply platformtour/platformtour.py]: Per F2F, let's still do this.

File: platformtour/public/js/src/platformtour_studio.js [hunk 34]
```diff
@@ -0,0 +1,63 @@
...
+    var buttonCancel = $element.find('.cancel-button');
+    var url = runtime.handlerUrl(element, 'studio_view_save');
+
+    var customSteps = $("#platformtour_studio_custom_steps");
+    var stepChoices = $("#platformtour_studio_enabled_steps");
+
+    var checkboxIconCheckedClass = "fa-check-square-o ";
+    var checkboxIconUncheckedClass = "fa-square-o ";
+
+    $(".step input", stepChoices).click(function() {
```
Comment (stvstnfrd) [inline platformtour/public/js/src/platformtour_studio.js]: Looks like we're still switching between single and double quotes..
Can you standardize on single for non-interpolated strings?

File: platformtour/public/js/src/platformtour_studio.js [hunk 35]
```diff
@@ -0,0 +1,63 @@
...
+    buttonCancel.on('click', function () {
+        runtime.notify('cancel', {});
+        return false;
+    });
+
+    buttonSave.on('click', function () {
+        var stepChoicesJson = [];
+        $(".step", stepChoices).each(function( index ) {
+            var stepChoiceKey = $(this).data("key");
+            var isFirstItem = stepChoiceKey === "intro";
```
Comment (stvstnfrd) [inline platformtour/public/js/src/platformtour_studio.js]: This is a little brittle; 
could you check if `index === 0`?

File: platformtour/templates/platformtour_studio.html (hunk 31 above)
Comment (stvstnfrd) [reply platformtour/templates/platformtour_studio.html]: It looks like for accessibility concerns, we could use the `aria-labelledby` attribute [1] and would no longer need the ID.
That said, we'd basically be duplicating the label's text and seeing as we already _do_ want the label for visual display, it's probably a bit of a stretch.

So if we need IDs for a11y purposes, let's just use it for a11y and not for styling and logic.

This should be as simple as adding a simple class name to the INPUTs and then performing look-ups as children;
see comments elsewhere.

[1] https://www.w3.org/WAI/tutorials/forms/labels/#using-aria-labelledby

Comment (stvstnfrd) [reply platformtour/templates/platformtour_studio.html]: Add `display-name` to class list.

File: platformtour/templates/platformtour_studio.html [hunk 36]
```diff
@@ -0,0 +1,61 @@
...
+            </div>
+            <span class="tip setting-help">This name appears in the horizontal navigation at the top of the page</span>
+        </li>
+
+         <li class="field comp-setting-entry is-set">
+            <div class="wrapper-comp-setting">
+                <label class="label setting-label" for="platformtour_studio_button_label">Button label</label>
+                <input class="input setting-input" id="platformtour_studio_button_label" value="{{ button_label }}" type="text">
```
Comment (stvstnfrd) [inline platformtour/templates/platformtour_studio.html]: Add `button-label` to class list.

File: platformtour/templates/platformtour_studio.html [hunk 37]
```diff
@@ -0,0 +1,61 @@
...
+            </div>
+            <span class="tip setting-help">This is the text that will appear on the button</span>
+        </li>
+
+         <li class="field comp-setting-entry is-set">
+            <div class="wrapper-comp-setting">
+                <label class="label setting-label" for="platformtour_studio_intro">Introduction text</label>
+                <input class="input setting-input" id="platformtour_studio_intro" value="{{ intro }}" type="text">
```
Comment (stvstnfrd) [inline platformtour/templates/platformtour_studio.html]: Add `intro` to class list.

File: platformtour/templates/platformtour_studio.html [hunk 38]
```diff
@@ -0,0 +1,61 @@
...
+            </div>
+            <span class="tip setting-help">This is the introduction that will precede the button and explain its presence to the user</span>
+        </li>
+
+        <li class="field comp-setting-entry is-set">
+            <div class="wrapper-comp-setting">
+                <label class="label setting-label" for="platformtour_studio_enabled_steps">Choose the steps for the Platform Tour</label>
+                <div id="platformtour_studio_enabled_steps">
```
Comment (stvstnfrd) [inline platformtour/templates/platformtour_studio.html]: Add `steps` to class list.

File: platformtour/templates/platformtour.html [hunk 39]
```diff
@@ -0,0 +1,13 @@
+<article>
+    <h2 class="problem-header">{{ display_name }}</h2>
+    <p>{{ intro }}</p>
+
+    <button class="navmaker">{{ button_label }}</button>
+
+    <div class="sr">
+        <h4>Screen Reader Accessible Version of the Platform Tour</h4>
+        <ol id="screen-reader-steps"></ol>
```
Comment (stvstnfrd) [inline platformtour/templates/platformtour.html]: Does this need to be an ID?
There can now be multiple instantiations of the xblock on a single page..

File: platformtour/public/css/platformtour_studio.css [hunk 40]
```diff
@@ -0,0 +1,38 @@
+#platformtour_studio_enabled_steps {
```
Comment (stvstnfrd) [inline platformtour/public/css/platformtour_studio.css]: `.xblock_platformtour_studio .steps`

File: platformtour/public/css/platformtour_studio.css [hunk 41]
```diff
@@ -0,0 +1,38 @@
...
+  display: inline-block;
+  width: 45%;
+  border: 1px solid #ccc;
+  vertical-align: top;
+}
+#platformtour_studio_enabled_steps .step {
```
Comment (stvstnfrd) [inline platformtour/public/css/platformtour_studio.css]: `.xblock_platformtour_studio .steps .step`

File: platformtour/templates/platformtour_studio.html [hunk 42]
```diff
@@ -0,0 +1,61 @@
...
+                    <div class="step {% if forloop.counter|divisibleby:2 %}even{% else %}odd{% endif %}" data-key="{{ step.key }}">
+                      <label>
+                        {{ step.title }}
+                      {% if not forloop.first %}
+                        <input type="checkbox" name="{{ step.key }}" value="{{ step.title }}" {% if step.enabled %}checked{% endif %} data-index="{{ forloop.counter0 }}">
+                        <i class="fa fa-lg {% if step.enabled %}fa-check-square-o{% else %}fa-square-o{% endif %}"></i>
+                      {% endif %}
+                      </label>
+                      {% if forloop.last %}
+                      <textarea class="field-data-control {% if step.enabled %}enabled{% endif %}" data-field-name="custom_steps" id="platformtour_studio_custom_steps" rows="5" cols="70">{{ custom_steps }}</textarea>
```
Comment (stvstnfrd) [inline platformtour/templates/platformtour_studio.html]: Add `custom-steps` to class list.

File: platformtour/public/css/platformtour_studio.css [hunk 43]
```diff
@@ -0,0 +1,38 @@
...
+#platformtour_studio_enabled_steps .step label input {
+  display: none;
+}
+#platformtour_studio_enabled_steps .step label i {
+  position: absolute;
+  right: 0;
+  width: 20px;
+}
+
+#platformtour_studio_custom_steps {
```
Comment (stvstnfrd) [inline platformtour/public/css/platformtour_studio.css]: `.xblock_platformtour_studio .custom-steps`

File: platformtour/public/css/platformtour_studio.css [hunk 44]
```diff
@@ -0,0 +1,38 @@
...
+  display: none;
+  width: 100%;
+  max-width: 100%;
+  box-sizing: border-box;
+  margin-top: 10px;
+}
+#platformtour_studio_custom_steps.enabled {
```
Comment (stvstnfrd) [inline platformtour/public/css/platformtour_studio.css]: `.xblock_platformtour_studio .custom-steps.enabled`

File: platformtour/public/css/platformtour_studio.css [hunk 45]
```diff
@@ -0,0 +1,38 @@
+#platformtour_studio_enabled_steps {
+  display: inline-block;
+  width: 45%;
+  border: 1px solid #ccc;
+  vertical-align: top;
+}
+#platformtour_studio_enabled_steps .step {
+  padding: 10px;
+}
+#platformtour_studio_enabled_steps .step.even {
```
Comment (stvstnfrd) [inline platformtour/public/css/platformtour_studio.css]: `.xblock_platformtour_studio .steps .step.even`

File: platformtour/public/css/platformtour_studio.css [hunk 46]
```diff
@@ -0,0 +1,38 @@
...
+  background-color: #ddd;
+}
+#platformtour_studio_enabled_steps .step label {
```
Comment (stvstnfrd) [inline platformtour/public/css/platformtour_studio.css]: `.xblock_platformtour_studio .steps .step label`

File: platformtour/public/css/platformtour_studio.css [hunk 47]
```diff
@@ -0,0 +1,38 @@
...
+  position: relative;
+  display: inline-block;
+  width: 100%;
+  font-size: 1em;
+  line-height: 1em;
+}
+#platformtour_studio_enabled_steps .step label input {
```
Comment (stvstnfrd) [inline platformtour/public/css/platformtour_studio.css]: `.xblock_platformtour_studio .steps .step label input`

File: platformtour/public/css/platformtour_studio.css [hunk 48]
```diff
@@ -0,0 +1,38 @@
...
+  display: none;
+}
+#platformtour_studio_enabled_steps .step label i {
```
Comment (stvstnfrd) [inline platformtour/public/css/platformtour_studio.css]: `.xblock_platformtour_studio .steps .step label i`

File: platformtour/public/js/src/platformtour_studio.js [hunk 49]
```diff
@@ -0,0 +1,63 @@
+function PlatformTourStudioUI(runtime, element) {
+    'use strict';
+
+    var $ = window.$;
+    var $element = $(element);
+    var buttonSave = $element.find('.save-button');
+    var buttonCancel = $element.find('.cancel-button');
+    var url = runtime.handlerUrl(element, 'studio_view_save');
+
+    var customSteps = $("#platformtour_studio_custom_steps");
```
Comment (stvstnfrd) [inline platformtour/public/js/src/platformtour_studio.js]: `$element.find('.custom-steps')`

File: platformtour/public/js/src/platformtour_studio.js [hunk 50]
```diff
@@ -0,0 +1,63 @@
...
+    var stepChoices = $("#platformtour_studio_enabled_steps");
```
Comment (stvstnfrd) [inline platformtour/public/js/src/platformtour_studio.js]: `$element.find('.steps')`

File: platformtour/public/js/src/platformtour_studio.js [hunk 51]
```diff
@@ -0,0 +1,63 @@
...
+            }
+        });
+        runtime.notify('save', {
+            message: 'Saving...',
+            state: 'start',
+        });
+        $.ajax(url, {
+            type: 'POST',
+            data: JSON.stringify({
+                'display_name': $("#platformtour_studio_display_name").val(),
```
Comment (stvstnfrd) [inline platformtour/public/js/src/platformtour_studio.js]: `$element.find('.display-name')`

File: platformtour/public/js/src/platformtour_studio.js [hunk 52]
```diff
@@ -0,0 +1,63 @@
...
+                'button_label': $("#platformtour_studio_button_label").val(),
```
Comment (stvstnfrd) [inline platformtour/public/js/src/platformtour_studio.js]: `$element.find('.button-label')`

File: platformtour/public/js/src/platformtour_studio.js [hunk 53]
```diff
@@ -0,0 +1,63 @@
...
+                'intro': $("#platformtour_studio_intro").val(),
```
Comment (stvstnfrd) [inline platformtour/public/js/src/platformtour_studio.js]: `$element.find('.intro')`

**Review by stvstnfrd — CHANGES_REQUESTED** [review]
Almost there..

File: platformtour/default_steps.py [hunk 54]
```diff
@@ -96,14 +113,17 @@
...
+                  ' question to the forum. Peers and course teams may be able to answer your'
+                  'question there.'),
         'position': 'left',
     },
     {
         'key': 'help_link',
         'title': 'Help Link',
         'element': 'a.doc-link',
-        'intro': 'For any technical issues or platform-specific questions, click on the \'Help\' link to access the Help Center or contact support.',
+        'intro': ('For any technical issues or platform-specific questions,'
```
Comment (stvstnfrd) [inline platformtour/default_steps.py]: Add a space to the end of this line, else it the next word is connected to the comma.

Comment (stvstnfrd) [reply platformtour/default_steps.py]: And while we're in there, let's clean up the formatting like:
```py
'intro': (
    'For any technical issues or platform-specific questions, '
)
```

File: platformtour/default_steps.py [hunk 55]
```diff
@@ -96,14 +113,17 @@
         'key': 'discussion_forum',
         'title': 'Discussion Forum',
         'element': '.course-tabs a[href*="/discussion"]',
-        'intro': 'For course-specific questions, click on the \'Discussion\' tab to post your question to the forum. Peers and course teams may be able to answer your question there.',
+        'intro': ('For course-specific questions, click on the \'Discussion\' tab to post your'
+                  ' question to the forum. Peers and course teams may be able to answer your'
```
Comment (stvstnfrd) [inline platformtour/default_steps.py]: Add a space to the end of this line, else it the next word is connected to the previous.

And while we're in there, let's clean up the formatting like:
```py
'intro': (
    'For any technical issues or platform-specific questions, '
)
```

File: platformtour/default_steps.py [hunk 56]
```diff
@@ -82,7 +96,10 @@
         'key': 'bookmarking',
         'title': 'Bookmarking',
         'element': '.bookmark-button-wrapper',
-        'intro': 'If you want to get back later to the content on a particular page, or you want to save it as something important, bookmark it. A Bookmarks folder on your course home page will contain a link to any page you bookmark for easy access later.',
+        'intro': ('If you want to get back later to the content on a particular page,'
```
Comment (stvstnfrd) [inline platformtour/default_steps.py]: Add a space to the end of this line, else it the next word is connected to the comma.

And while we're in there, let's clean up the formatting like:
```py
'intro': (
    'For any technical issues or platform-specific questions, '
)
```

Comment (stvstnfrd) [reply platformtour/default_steps.py]: Note: this comment applies to all of the changes to these multiline strings.

File: platformtour/platformtour.py [hunk 57]
```diff
@@ -202,9 +202,21 @@ def workbench_scenarios():
              """),
             ("Multiple PlatformTourXBlock",
              """<vertical_demo>
-                <platformtour/>
-                <platformtour/>
-                <platformtour/>
+                    <platformtour
+                        display_name="Platform Tour 1"
+                        button_label="Start  Tour #1"
```
Comment (stvstnfrd) [inline platformtour/platformtour.py]: Are there 2 spaces here on purpose?

**Review by stvstnfrd — CHANGES_REQUESTED** [review]

**Review by stvstnfrd — APPROVED** [review]

**Pull request merged**